File: trab2.c
#include <stdio.h>
#include <math.h>
#define MAX_ITERATIONS 100
double function(double x){
	double cs=9.0;
	double c0=2.0;
	double ka=0.88;
	double t=x;

	return cs-(cs-c0)*exp(-ka*t);
}

double function_zero(double t){
	return 1 - 7*exp(-0.88*t);
  //return 1.94591014906-0.88*t;//9-(9-2)*e^(-0.88t) -> ln(1-7*e^(-0.88*t))=0 -> ln(7) + (-0.88t)=0
}

double derivative_zero(double t){
  return 0.88*7*exp(-0.88*t);
  //return -0.88;
}
 	
double derivative(double x){
	double cs=9.0;
	double c0=2.0;
	double ka=0.88;
	double t=x;

	return (ka)*(cs-c0)*exp(-ka*t);
}

double newton_raphson_core(double x0){
	double x1=0;
    x1=x0 - function_zero(x0)/derivative_zero(x0);
	return x1;	
}

double newton_raphson(double x0, double tolerance){
	double x1=0;
	double error=tolerance * 2;	
	int iteration=0;
    
	while( error > tolerance && iteration < MAX_ITERATIONS ){
		x1=newton_raphson_core(x0);
		error=fabs(x1-x0);
        printf("x0:%lf. x1:%lf. erro:%lf.funcao %lf\n",x0,x1,error,function(x1));
		x0=x1;
		iteration++;
	}
	
	if(iteration >= MAX_ITERATIONS){
	  printf("Não ficou dentro da tolerância de erro em %d iterações.\n", MAX_ITERATIONS);
    }
    
    return x1;
}

int main(int argc, char** argv){

  printf("\nt = 1 dia\n");
  printf("resultado = %lf\n\n", newton_raphson(1.0, 10E-6));
  printf("t = 5 dias\n");
  printf("resultado = %lf\n\n", newton_raphson(5.0, 10E-6));
  printf("t = 10 dias\n");
  printf("resultado = %lf\n\n", newton_raphson(10.0, 10E-6));
	
}

File: trab2.py
# coding=utf-8
import math

MAX_ITERATIONS = 100

def function(t):
    cs = 9.0
    c0 = 2.0
    ka = 0.88

    return cs - (cs - c0) * math.exp(-ka * t)


def derivative(t):
    cs = 9.0
    c0 = 2.0
    ka = 0.88

    return ka * (cs - c0) * math.exp(-ka * t)


def function_zero(t):
    return 1 - 7 * math.exp(-0.88 * t)


def derivative_zero(t):
    return 0.88 * 7 * math.exp(-0.88 * t)


def newton_raphson_core(x0):
    return x0 - function(x0)/derivative(x0)


def newton_raphson(x0, tolerance):
    x1 = 0
    error = tolerance + 1
    iteration = 0

    try:
        while error > tolerance and iteration < MAX_ITERATIONS:
            x1 = newton_raphson_core(x0)
            error = math.fabs(x1 - x0)
            #print("x0:%lf. x1:%lf. erro:%lf.funcao %lf\n", x0, x1, error, function(x1))
            x0 = x1
            iteration = iteration + 1

        if iteration >= MAX_ITERATIONS:
            print("Nao ficou dentro da tolerancia de erro em " + str(MAX_ITERATIONS) + "iteracoes.")
    except OverflowError:
        print("Erro: Overflow, nao foi possível convergir.")
    except:
        print("Erro! Nao foi possível convergir.")

    return x1


print("\nt = 1 dia")
print("resultado =", newton_raphson(1.0, 10E-6))
print("")
print("t = 5 dias")
print("resultado =", newton_raphson(5.0, 10E-6))
print("")
print("t = 10 dias")
print("resultado =", newton_raphson(10.0, 10E-6))
